File: src/00030_SearchInRotatedSortedArray.py
import math

class Solution:
    def search(self, nums: list[int], target: int) -> int:
        if len(nums) == 0:
            return -1

        if len(nums) == 1:
            if nums[0] == target:
                return 0
            else:
                return -1 

        pivot = searchPivot(nums, 0, len(nums)-1)
        first = searchBinary(nums, 0, pivot, target)
        if (first == -1):
            return searchBinary(nums, pivot+1, len(nums)-1, target)
        else:
            return first

def searchPivot(nums: list[int], start, end) -> int:
    if nums[start] <= nums[end]:
        return start-1
    else:
        check = math.floor((end+start)/2)
        if (nums[start] <= nums[check]):
            return searchPivot(nums, check+1, end)
        else:
            return searchPivot(nums, start, check)

def searchBinary(nums: list[int], left, right, target) -> int:
    while(left <= right):
        check = math.floor((right+left)/2)

        if (target == nums[check]):
            return check

        if (target < nums[check]):
            right = check - 1
        else:
            left = check + 1

    return -1

File: src/C#/00030_SearchInRotatedSortedArray.cs
public class Solution
{
    public int Search(int[] nums, int target)
    {
        // find the pivot point (must be log N)
        // search number is correspond range (nums[k] -> nums[n-1]) (nums[0] -> nums[k-1])
        // search by binary search (logN)

        if (nums.Length == 0)
        {
            return -1;
        }

        if (nums.Length == 1)
        {
            return nums[0] == target ? 0 : -1;
        }

        var pivotIndex = GetPivotPointIndex(nums, 0, nums.Length - 1);
        if (pivotIndex != 0)
        {
            if (nums[nums.Length - 1] < target)
            {
                return BinarySearch(nums, 0, pivotIndex - 1, target);
            }
            else
            {
                return BinarySearch(nums, pivotIndex, nums.Length - 1, target);
            }
        }
        else
        {
            return BinarySearch(nums, 0, nums.Length - 1, target);
        }
    }

    public int GetPivotPointIndex(int[] nums, int start, int end)
    {
        int middle = (start + end) / 2;
        if (nums[start] <= nums[middle] && nums[middle] <= nums[end])
        {
            return start;
        }

        if (nums[start] > nums[middle])
        {
            return GetPivotPointIndex(nums, start, middle);
        }

        if (nums[middle] > nums[end])
        {
            return GetPivotPointIndex(nums, middle+1, end);
        }

        return 0;
    }

    public int BinarySearch(int[] nums, int start, int end, int target)
    {
        int middle = (start + end) / 2;

        if (middle == start && middle == end)
        {
            if (nums[middle] == target)
            {
                return middle;
            }
            else
            {
                return -1;
            }
        }

        if (nums[middle] >= target)
        {
            return BinarySearch(nums, start, middle, target);
        }
        else
        {
            return BinarySearch(nums, middle+1, end, target);
        }
    }
}